source/host_resolver.c

```diff
@@ -520,16 +520,22 @@ static inline void process_records(
     AWS_LOGF_TRACE(AWS_LS_IO_DNS, "static: remaining record count for host %d", (int)record_count);
 
     /* if we don't have any known good addresses, take the least recently used, but not expired address with a history
-     * of spotty behavior and upgrade it for reuse. If it's expired, leave it and let the resolve fail. Better to fail
-     * than accidentally give a kids' app an IP address to somebody's adult website when the IP address gets rebound to
-     * a different endpoint. The moral of the story here is to not disable SSL verification! */
-    if (!record_count) {
-        size_t failed_count = aws_cache_get_element_count(failed_records);
-        for (size_t index = 0; index < failed_count; ++index) {
-            struct aws_host_address_cache_entry *lru_element_entry = aws_lru_cache_use_lru_element(failed_records);
-            if (timestamp >= lru_element_entry->address.expiry) {
-                continue;
-            }
+     * of spotty behavior and upgrade it for reuse. */
+    bool should_promote = (record_count == 0);
+    size_t failed_count = aws_cache_get_element_count(failed_records);
+    for (size_t index = 0; index < failed_count; ++index) {
+        struct aws_host_address_cache_entry *lru_element_entry = aws_lru_cache_use_lru_element(failed_records);
+
+        /* remove expired entries */
+        if (timestamp >= lru_element_entry->address.expiry) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_IO_DNS,
+                "static: purging expired bad record %s for %s",
+                lru_element_entry->address.address->bytes,
+                lru_element_entry->address.host->bytes);
+            aws_cache_remove(failed_records, lru_element_entry->address.address);
```

Comment:
is removing elements while iterating through cache safe here? a bit of naive question cause im not entirely familiar with cache impl